graylog2-web-interface/src/views/components/QueryTabs.jsx

```diff
@@ -1,64 +1,112 @@
-        {tabs}
-      </Tabs>
-    </span>
-  );
-};
+class QueryTabs extends React.Component<Props> {
+  static propTypes = {
+    children: PropTypes.node,
+    onRemove: PropTypes.func.isRequired,
+    onSaveAsView: PropTypes.func.isRequired,
+    onSaveView: PropTypes.func.isRequired,
+    onSelect: PropTypes.func.isRequired,
+    onTitleChange: PropTypes.func.isRequired,
+    queries: PropTypes.object.isRequired,
+    selectedQueryId: PropTypes.string.isRequired,
+    titles: PropTypes.object.isRequired,
+  }
+
+  static defaultProps = {
+    children: null,
+  }
 
-QueryTabs.propTypes = {
-  children: PropTypes.node,
-  onSaveView: PropTypes.func.isRequired,
-  onSaveAsView: PropTypes.func.isRequired,
-  onSelect: PropTypes.func.isRequired,
-  onRemove: PropTypes.func.isRequired,
-  onTitleChange: PropTypes.func.isRequired,
-  queries: PropTypes.object.isRequired,
-  selectedQuery: PropTypes.string.isRequired,
-  titles: PropTypes.object.isRequired,
-};
+  queryTitleEditModal: ?QueryTitleEditModal
 
-QueryTabs.defaultProps = {
-  children: null,
-};
+  openTitleEditModal = (activeQueryTitle: string) => {
+    if (this.queryTitleEditModal) {
+      this.queryTitleEditModal.open(activeQueryTitle);
+    }
+  }
+
+  render() {
+    const {
+      children,
+      onRemove,
+      onSaveAsView,
+      onSaveView,
+      onSelect,
+      onTitleChange,
+      queries,
+      selectedQueryId,
+      titles,
+    } = this.props;
+    const queryTitles = titles;
+    const queryTabs = queries.map((id, index) => {
+      const title = queryTitles.get(id, `Query#${index + 1}`);
+      const tabTitle = (
+        <QueryTitle active={id === selectedQueryId}
+                    id={id}
+                    onClose={() => onRemove(id)}
+                    openEditModal={this.openTitleEditModal}
+                    title={title} />
+      );
+      return (
+        <Tab eventKey={id}
+             key={id}
+             mountOnEnter
+             title={tabTitle}
+             unmountOnExit>
+          {children}
+        </Tab>
+      );
+    });
+    const newTab = <Tab key="new" eventKey="new" title="+" />;
+    const tabs = [queryTabs, newTab];
+
+    return (
+      <span>
+        <span className="pull-right">
+          <ViewActionsMenu onSaveView={onSaveView} onSaveAsView={onSaveAsView} />
+        </span>
+        <Tabs activeKey={selectedQueryId}
+              animation={false}
+              id="QueryTabs"
+              onSelect={onSelect}>
+          {tabs}
+        </Tabs>
+        {/*
+          The title edit modal can't be part of the QueryTitle component,
+          due to the react bootstrap tabs keybindings.
+          The input would always lose the focus when using the arrow keys.
+        */}
+        <QueryTitleEditModal onTitleChange={(newTitle: string) => onTitleChange(selectedQueryId, newTitle)}
+                             ref={(queryTitleEditModal) => { this.queryTitleEditModal = queryTitleEditModal; }} />
```

Comment:
Normally I would have placed this component inside the QueryTitle component. But due to the described problems I thought this is the best idea. Even now where the input is inside the modal the key binding was very strange. I was able to prevent it from creating a new query tab when using the arrow keys, but the input still lost the focus. That was very annoying when you want to e.g. select the full text